File: problem_set_1/palindrome_pairs.js
// A function for validating a string, returns a boolean
function isPalindrome(word) {
	for (let i = 0; i < word.length / 2; i++) {
		//Compares the characters starting from the outermost part of the word up to the center.
		if (word[i] !== word[word.length - 1 - i]) return false;
	}

	return true;
}

// A function for finding palindrome pairs, returns a nested array
function palindromePairs(words) {
	let result = [];

	for (let i = 0; i < words.length; i++) {
		//Start j with i + 1 to prevent comparing a word with itself
		for (let j = i + 1; j < words.length; j++) {
			const first = words[i] + words[j];
			const second = words[j] + words[i];

			// Check first combined word
			if (isPalindrome(first)) {
				result.push([i, j]);
			}

			// Check second combined word
			if (isPalindrome(second)) {
				result.push([j, i]);
			}
		}
	}

	return result;
}

// Main function
function main() {
	// let words = ["bat", "tab", "cat"];
	// console.log("Input: " + JSON.stringify(words));
	// // Check input
	// const result = palindromePairs(words);
	// console.log("Output: " + JSON.stringify(result));

	// - - - - - - - FOR MANUAL INPUT - - - - - - - //
	const readline = require("node:readline");
	const rl = readline.createInterface({
		input: process.stdin,
		output: process.stdout,
	});

	// Use a Set to store unique words
	let words = new Set();

	rl.question("Enter a list of words: ", (input) => {
		if (input.trim().length !== 0) {
			// Split the input and add unique words to the Set
			input.split(/\s+/).forEach((word) => {
				words.add(word);
			});

			// Convert Set back to an array
			words = Array.from(words);

			// Check input
			const result = palindromePairs(words);
			console.log("Output: " + JSON.stringify(result));
		}
		rl.close();
	});
}

// Run the program
main();

File: problem_set_1/palindrome_pairs.cs
namespace Problem.Set.One {
    using System;
    using System.Linq;
    using System.Collections.Generic;

    class PalindromePairs {
        // A function for validating a string, returns a boolean
        public static bool isPalindrome(string word) {
            for (int i = 0; i < word.Length / 2; i++) {
                if (word[i] != word[word.Length - 1 - i]) return false;
            }

            return true;
        }

        // A function for finding palindrome pairs, returns a nested array
        public static int[][] palindromePairs(string[] words) {
            List<int[]> result = new List<int[]>();

            for (int i = 0; i < words.Length; i++) {
                for (int j = i + 1; j < words.Length; j++) {
                    string first = words[i] + words[j];
                    string second = words[j] + words[i];

                    // Check first combined word
                    if (isPalindrome(first)) {
                        result.Add(new int[] { i, j });
                    }

                    // Check second combined word
                    if (isPalindrome(second)) {
                        result.Add(new int[] { j, i });
                    }
                }
            }

            return result.ToArray();
        }

        // A function for displaying output
        static void displayResult(int[][] result) {
            // Format the output as nested arrays
            Console.Write("Output: [");
            for (int i = 0; i < result.Length; i++)
            {
                Console.Write("[");
                Console.Write(string.Join(", ", result[i]));
                Console.Write("]");
                if (i < result.Length - 1)
                {
                    Console.Write(",");
                }
            }
            Console.WriteLine("]");
        }

        // Main function
        static void Main(string[] args) {
            // string[] words = new string[]  { "bat", "tab", "cat" };
            // string formattedArray = "[" + string.Join(", ", words) + "]";
            // Console.WriteLine("Input: " + formattedArray);

            // int result = lengthOfLIS(numbers);
            // Console.WriteLine("Output: " + result);
            // displayResult(result);

            // - - - - - - - FOR MANUAL INPUT - - - - - - - //
            Console.WriteLine("Input: ");
            string input = Console.ReadLine().Trim();

            //Proceed only if input is not empty.
            if (!string.IsNullOrWhiteSpace(input)) {
                //Split the input based on space
                string[] words = input.Split(
                    new char[] { ' ' }, 
                    StringSplitOptions.RemoveEmptyEntries);

                // Check input
                int[][] result = palindromePairs(words);
                displayResult(result);
            }

            //Prevent the screen from closing until a key is pressed.
            Console.WriteLine("Press any key to exit...");
            Console.ReadKey();
        }
    }
}
